tools/eslint-plugin/dist/rules/no-internal-barrel-imports.js

```diff
@@ -0,0 +1,297 @@
+        type: "object",
+        additionalProperties: false,
+        properties: {
+          /**
+           * This option may be replaced in the future with "required-barrel-modules", which
+           * indicates that all modules importing a symbol re-exported by that barrel module
+           * must import it from the barrel module, and from nowhere else. In some complicated dependency situations,
+           * forcing usage of a barrel module will enforce a particular dependency order. Modules that are imported by the barrel
+           * file will not need to import sibling modules from the barrel.
+           */
+          [OPTION_IGNORED_BARREL_MODULES]: {
+            type: "array",
+            description: "Usage of re-exports from these paths are ignored.",
+            default: [],
+          },
+        },
+      },
+    ],
+    fixable: "code",
+    messages,
+  },
+
+  create(context) {
+    const parserServices = getParserServices(context);
+    /** @type {ts.Program} */
+    const program = parserServices.program;
+    const checker = program.getTypeChecker();
+    const maybeTsConfig = program.getCompilerOptions().configFilePath;
+
+    const extraOpts = context.options[0];
+    const ignoredBarrelModules = (
+      (maybeTsConfig &&
+        extraOpts &&
+        extraOpts[OPTION_IGNORED_BARREL_MODULES]) ||
+      []
+    ).map((p) => {
+      /** @type {string} earlier check means it is definitely a string */
+      const tsConfig = maybeTsConfig;
+      return OsPaths.normalize(OsPaths.resolve(OsPaths.dirname(tsConfig), p));
+    });
+
+    return {
+      ImportDeclaration(node) {
+        /** @param {ts.Symbol | undefined} symbol */
+        function getRelativeImportPathForExportedSymbol(symbol) {
+          if (symbol === undefined) return "";
+          const declaration = symbol.valueDeclaration || symbol.declarations[0];
+          const fileOfExport = declaration.getSourceFile();
+          const path = TsImportPaths.normalize(
+            withoutExt(
+              TsImportPaths.relative(
+                TsImportPaths.dirname(thisModule.fileName),
+                fileOfExport.fileName
+              )
+            )
+          );
+          // a path unprefixed by "./" is interpretted as from node_modules
+          const ensurePrefixed = (path) =>
+            path[0] === "." ? path : `./${path}`;
+          return ensurePrefixed(path);
+        }
+
+        if (typeof node.source.value !== "string")
+          throw Error("Invalid input source");
+
+        /** @type{ts.ImportDeclaration} */
+        const importNodeTs = parserServices.esTreeNodeToTSNodeMap.get(node);
+        if (!importNodeTs)
+          throw Error("equivalent typescript node could not be found");
+
+        if (importNodeTs.importClause && importNodeTs.importClause.isTypeOnly)
+          return;
+
+        const thisModule = importNodeTs.getSourceFile();
+
+        if (!thisModule.resolvedModules.has(importNodeTs.moduleSpecifier.text))
```

Comment:
resolvedModules is an internal API, consider using ts.getResolvedModule instead

Reply:
I've setup jsconfig.json and JSDoc comments for type checking in this file, and even so, `ts.getResolvedModule` is not found in typescript@4.3.5. I haven't been finding it easy to navigate the typescript API, how do you know getResolvedModule is not internal?

Reply:
yeah, navigating their API is really irritating, and it has very little documentation. I just have the Typescript compiler's source cloned and I check things there if I need to. I saw the red squiggle, found the definition of `resolvedModules` and there's a comment above it:
```ts
// Content of this field should never be used directly - use getResolvedModuleFileName/setResolvedModuleFileName functions instead
```
`getResolvedModuleFileName` doesn't exist but `getResolvedModule` does, so I just assumed the comment is outdated and you should use the latter one.

Reply:
I now see the internal ones are marked `@internal`, which is what I'd expect but didn't see before (maybe they use the `--stripInternal` compiler option, I just started reading the JS output). Not sure why `getResolvedModule` isn't found on `ts`

Reply:
cloning typescript is probably a good idea, I've been searching in github, but I've always disliked github's search personally...